pkg/obfuscate/appsec.go

```diff
@@ -0,0 +1,538 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package obfuscate
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+// ObfuscateAppSec obfuscates the given appsec tag value in order to remove
+// sensitive values from the appsec security events.
+func (o *Obfuscator) ObfuscateAppSec(val string) string {
+	keyRE := o.opts.AppSec.ParameterKeyRegexp
+	valueRE := o.opts.AppSec.ParameterValueRegexp
+	if keyRE == nil && valueRE == nil {
+		return val
+	}
+	ob := appsecEventsObfuscator{
+		keyRE:   keyRE,
+		valueRE: valueRE,
+	}
+	output, err := ob.obfuscate(val)
+	if err != nil {
+		o.log.Errorf("unexpected error while obfuscating the appsec events: %v", err)
+		return val
+	}
+	return output
+}
+
+type appsecEventsObfuscator struct {
+	keyRE, valueRE *regexp.Regexp
+}
+
+type unexpectedScannerOpError int
+
+func (err unexpectedScannerOpError) Error() string {
+	return fmt.Sprintf("unexpected json scanner operation %d", err)
+}
+
+// inputDiff holds the list of diffs, built by the obfuscator, and to ultimately
+// apply to the input.
+type inputDiff []struct {
+	from, to int
+	value    string
+}
+
+// Add a new entry to the list. Note that due to how the apply() method works
+// by creating the output string by walking this list, the list entries must be
+// sorted in ascending "input offset" order.
+func (d *inputDiff) add(from, to int, value string) {
+	elt := struct {
+		from  int
+		to    int
+		value string
+	}{
+		from:  from,
+		to:    to,
+		value: value,
+	}
+	diff := *d
+	l := len(diff)
+	i := sort.Search(l, func(i int) bool {
```

Comment:
This is the part where it makes sense to start leaving comments.
```suggestion
        // Use binary search to find the index for the new entry.
        // We want the list sorted incrementally.
	i := sort.Search(l, func(i int) bool {
```

Reply:
Done in https://github.com/DataDog/datadog-agent/pull/11324/commits/48234cfc628802b9faf4ed9289d9742eaf5dc83d